### PR TITLE
New App: Tides & Temp

### PR DESCRIPTION
**Pull request by sudeepban**
# Description
Tides & Temp app for displaying predicted tides from NOAA tide stations and water temperature from NDBC buoy stations. See https://tidesandcurrents.noaa.gov/tide_predictions.html for NOAA tide station IDs. See https://www.ndbc.noaa.gov/obs.shtml for NDBC buoy station IDs. Wave graphic moves up and down based on estimated current tide calculated from the previous and next low / high tides.

tides_and_temp

# Copilot

### Summary
🌊🌡️📄

This pull request introduces a new applet for tidbyt that shows tide and water temperature data from NOAA and NDBC stations. The applet is implemented in `tides_and_temp.star` and uses render widgets and a schema object. The applet is also registered in the `manifest.yaml` file with its metadata.

> _`Tides & Temp` app_
> _Shows the sea's mood and warmth_
> _Manifest for spring_

### Walkthrough
*  Add the Tides & Temp applet, which displays predicted tides and water temperature from NOAA and NDBC stations ([link](https://github.com/tidbyt/community/pull/1506/files?diff=unified&w=0#diff-3cbc1c97c6acd593129b4b6ef0055bdae1d7e25000c5ef1831eaecd594bb7fccR1-R8), [link](https://github.com/tidbyt/community/pull/1506/files?diff=unified&w=0#diff-eea0cf481e8f71702f3c2d4583034be91ee9c98cbaa419a8e9f75bf29f1ada86R1-R202))

### CODE REVIEW

**Comment by sudeepban** [comment]
I have read the CLA Document and I hereby sign the CLA

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [inline apps/tidesandtemp/tides_and_temp.star:102]: Given the size of this block, I'm thinking it might be a good idea to scale the graphic down ahead of time?

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [inline apps/tidesandtemp/tides_and_temp.star:158]: A common pattern for getting timezone data is to use `schema.Location`. That allows the user to select a specific location, and gives you the associated timezone in the location object.

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [inline apps/tidesandtemp/tides_and_temp.star:42]: Would it be appropriate to add a `ttl_seconds` flag to this call? Not sure if they provide cache-control headers?

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [reply apps/tidesandtemp/tides_and_temp.star:158]: Also, you automatically get the device's configured location in the magic '$tz' config variable. Not sure if that's good enough or if it's important for this to be fully configurable.

**Review by matslina — DISMISSED** [review]

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:42]: I didn't see anything at TTL or cache-control headers at https://api.tidesandcurrents.noaa.gov/api/prod/ - let me know if you see anything relevant I could incorporate

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:158]: I thought about using the `schema.Location` option or the magic `$tz` config variable, but I thought it may be advantageous for the user to have more direct control over setting the timezone.

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:102]: Could you elaborate?  I also thought the size of the string was a little out of control but wasn't sure what I could do to improve it. It's already a small image, only 6x15, but it does have a decent number of frames to tie out the different waves so it cycles nicely.

Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:102]: I imagine one option would be to get the animated GIF properly hosted somewhere so I could reference that way, but wasn't sure if Tidbyt offers any such option.

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [reply apps/tidesandtemp/tides_and_temp.star:158]: Got it. In that case I still think `Location` is a good option to consider. Gives you handling of e.g. daylight's savings for free, and is likely more intuitive for the user (most people don't think much about their timezone's current offset from UTC for some reason) =). But it's your call!

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [reply apps/tidesandtemp/tides_and_temp.star:102]: When I decode that base64 block, I get a 1590x636 GIF. It looks like it's a 15x6 scaled up by a factor of 106.

Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:102]: Ohh interesting, let me take a closer look at that, thanks

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:158]: I think you're right, I will switch to using Location

Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star]: Switched to using `schema.Location` now

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star]: I tried rescaling to 15x6, and it actually made the animated GIF look not quite right. Rescaling to 45x18 looked fine, so it still reduces the size of the base64 block substantially.

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [reply apps/tidesandtemp/tides_and_temp.star:44]: The simplest thing is probably for you to decide how long the data retrieved is good for. So for instance, if it's ok to use the buoy data for (say) 1 hour, you could do `http.get(NDBC_BUOY_URL.format(station_id = ndbcBuoyStationID), ttl_seconds = 3600)`. This means subsequent requests for the exact same URL will not hit the NOAA server, but will instead be served from our cache. After 3600 seconds (=1h) the cached data will get thrown out and the next request will refresh the data from the API again.

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star]: Great, thanks

File: apps/tidesandtemp/tides_and_temp.star
Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star:44]: I see, that makes sense. My preference would be for it to be up to date, with the latest water temperature reading as soon as it's made available. It seems like adding any non-zero `ttl_seconds` would introduce some potential arbitrary delay.

**Comment by sudeepban** [comment]
@matslina is this change good to merge at this point?  Let me know if there's any other feedback I can address, thanks for your review!

File: apps/tidesandtemp/tides_and_temp.star
Comment (matslina) [reply apps/tidesandtemp/tides_and_temp.star:44]: Totally understand wanting up to date data. That makes sense. Unfortunately we do have to cache responses to keep request volumes at a resonable level.

I took a look at the data served on the buoy url, and it includes cache-control headers requesting 600s caching of responses, so that's the minimum we should do there. In fact, that's what'll happen automatically when `ttl_seconds` is left out.

There doesn't seem to be such a header on the other endpoint, but if I understand this correctly, it only serves tide predictions. These likely do not change much in the short term, so an hour of caching shouldn't hurt I imagine?

For reference, the existing apps nooatides and nooabuoy cache responses for 4 hours and 10 minute respectively.

Comment (sudeepban) [reply apps/tidesandtemp/tides_and_temp.star]: Thanks for the details.  I accepted your proposed changes, this should be fine!

**Comment by sudeepban** [comment]
@matslina is there any other feedback to address or can this be merged now?

**Comment by sudeepban** [comment]
> @matslina is there any other feedback to address or can this be merged now?

@betterengineering @matslina anything I can do to move this along? Really wanted to make this available via the mobile app for the beach season!

**Comment by wilbah** [comment]
I would like to see / use this!

**Comment by matslina** [comment]
Just got back from vacation. Sorry for the delay!

**Comment by sudeepban** [comment]
> Just got back from vacation. Sorry for the delay!

Ahh no problem, thank you! Hope you had a great vacation

**Comment by sudeepban** [comment]
@matslina I was curious, what is the process that makes a new app like this available via the Tidbyt mobile app?

**Comment by sudeepban** [comment]
> When we merge a PR, CI takes over and releases the app. It should be available in the mobile app within minutes. It won't get categorized automatically, i.e. you won't see it under "what's new", but it should be avaialable when you search.

That's awesome, I see it now! Thank you

**Pull request merged**
